### PR TITLE
Solidify config

### LINKED ISSUE #69: Dynamic Properties are deprecated in PHP 8.2
I'm using Debian Bookworm with icinga-php-library 0.10.1-1 and PHP 8.2.  When I log in in icingaweb2, I see three times this message in the left sidebar between the search bar in the menu:
`Deprecated: Creation of dynamic property ipl\Sql\Config::$skip_validation is deprecated in /usr/share/icinga-php/ipl/vendor/ipl/sql/src/Config.php on line 32`

### PR DESCRIPTION
**Pull request by TAINCER**
- Fixes PHP 8.2 dynamic properties deprecation warnings
 - Converts kes to camelCase

closes #67
fixes #69

### CODE REVIEW

File: src/Config.php [hunk 1]
```diff
@@ -62,11 +44,45 @@ public function __construct($config)
...
      * connection.
      *
      * @var array
      */
     public $options;
+
+    /**
+     * Create a new SQL connection configuration from the given configuration key-value pairs
+     *
+     * The given configuration is constraint to the following key-values:
```
Comment (lippserd) [inline src/Config.php:57]: The given configuration is constraint to the class' public properties.

File: src/Config.php [hunk 2]
```diff
@@ -62,11 +44,45 @@ public function __construct($config)
...
+     *     db, host, port, dbname, username, password, use_ssl, charset, and options
+     *
+     * Keys will be converted to camelcase, e.g. `use_ssl -> useSsl`, `UserName -> username`
```
Comment (lippserd) [inline src/Config.php:60]: Keys will be converted to camelCase, e.g. `use_ssl` → `useSsl`.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: src/Config.php [hunk 3]
```diff
@@ -62,11 +44,41 @@ public function __construct($config)
...
+                '%s expects parameter one to be iterable, got %s instead',
+                __METHOD__,
+                get_php_type($config)
+            ));
+        }
+
+        foreach ($config as $key => $value) {
+            $camelKey = Str::camel($key);
+
+            if (!property_exists(static::class, $camelKey)) {
```
Comment (lippserd) [inline src/Config.php:77]: Space missing after `!`.

**Review by lippserd — CHANGES_REQUESTED** [review]
I would rephrase the second commit to the following:

> Config: Constrain and camelCasify keys

Also the commit description should explain why this change is necessary and its benefits, e.g.
* It fixes dynamic properties error for PHP 8.2+. Also explain how this is happening. Remember the `resources.ini` topic we've talked about.
* camelCase for properties because that's our code style.
* Allows mixing upper and lowercase which makes it bulletproof to be used with any input.

The PR also lacks a proper description. I would add the consideration from above.

**Comment by lippserd** [comment]
Please also mention that the PR fixes issue 69.

File: src/Config.php (hunk 3 above)
Comment (lippserd) [reply src/Config.php:77]: Can you please check whether adding `Generic.Formatting.SpaceAfterNot` to our phpcs.xml would help to let our CI fail in such cases?

File: phpcs.xml [hunk 4]
```diff
@@ -15,6 +15,10 @@
         <exclude name="PSR12.Properties.ConstantVisibility.NotFound"/>
     </rule>
 
+    <rule ref="Generic.Formatting.SpaceAfterNot">
```
Comment (lippserd) [inline phpcs.xml:18]: Please remove that change from here.

File: src/Config.php [hunk 5]
```diff
@@ -62,11 +44,41 @@ public function __construct($config)
 
     /**
      * PDO connect options
-     *
```
Comment (lippserd) [inline src/Config.php:65]: Please keep the newline.

**Review by lippserd — CHANGES_REQUESTED** [review]

**Review by lippserd — APPROVED** [review]

File: tests/ConfigTest.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            'uSe_SsL' => true
+        ]);
+
+        $this->assertSame($config->db, 'mysql');
+        $this->assertSame($config->host, 'localhost');
+        $this->assertSame($config->useSsl, true);
+        $this->assertSame($config->dbname, 'test_dbname');
+    }
+
+    public function testAllPropertiesCanBeSet(): void
```
Comment (nilmerg) [inline tests/ConfigTest.php:35]: This misses the ssl properties like `ssl_key`.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Comment by nilmerg** [comment]
And remove the `Closes` from the description.

File: src/Config.php [hunk 7]
```diff
@@ -51,6 +30,27 @@ public function __construct($config)
     /** @var string Password to use for authentication */
     public $password;
 
+    /** @var bool Whether to use SSL */
+    public $useSsl;
```
Comment (lippserd) [inline src/Config.php:34]: Unfortunately, the SSL options are specific to the database driver used, e.g. MySQL, PostgreSQL, MSSQL all have different options. So we cannot simply define them here. We could do the following:

1. Store options that are not represented by public properties in an extra array and implement `__get`, `__set`. The `property_exists()` call can be removed then.
2. Concrete `Config` implementations for the database drivers. This would require quite some rework so I opt for option 1.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: src/Config.php (hunk 7 above)
Comment (TAINCER) [reply src/Config.php:34]: I've kept the `property_exists()` call to determine, if the values should be saved in the dedicated property, or the array.

File: src/Config.php [hunk 8]
```diff
@@ -51,6 +30,15 @@ public function __construct($config)
...
+
+    /** @var array Dynamic options for SQL SSL connections, etc. */
+    public $dynamicOptions = [];
```
Comment (lippserd) [inline src/Config.php:37]: Should not be public and renamed to something like `extraSettings` or `specificSettings`.

File: src/Config.php [hunk 9]
```diff
@@ -69,4 +57,59 @@ public function __construct($config)
...
+        }
+
+        return null;
+    }
+
+    public function __set(string $name, $value): void
+    {
+        $this->setProperty($name, $value);
+    }
+
```
Comment (lippserd) [inline src/Config.php:104]: Please also implement `__isset()`.

File: src/Config.php [hunk 10]
```diff
@@ -51,6 +30,15 @@ public function __construct($config)
...
     public $password;
 
+    /** @var bool Whether to use SSL */
+    public $useSsl;
+
+    /** @var array Dynamic options for SQL SSL connections, etc. */
+    public $dynamicOptions = [];
+
+    /** @var bool Whether to verify the SSL Certification */
+    public $sslDoNotVerifyServerCert;
```
Comment (lippserd) [inline src/Config.php:40]: Should not be a public property as it is specific to the database driver used.

File: src/Config.php (hunk 7 above)
Comment (lippserd) [reply src/Config.php:34]: Also, `useSsl` should not be a public property as it is specific to the database driver used.

File: src/Config.php [hunk 11]
```diff
@@ -69,4 +51,79 @@ public function __construct($config)
...
+            if ($name === $key) {
+                return $this->$name;
+            }
+        }
+
+        if (array_key_exists($name, $this->extraSettings)) {
+            return $this->extraSettings[$name];
+        }
+
+        return null;
```
Comment (lippserd) [inline src/Config.php:104]: I would throw an exception to protect against unexpected results:

`throw new OutOfRangeException(sprintf('Property %s does not exist', $name));`

**Review by lippserd — CHANGES_REQUESTED** [review]

**Comment by lippserd** [comment]
I'm closing this PR and creating a new one with me as the author, because honestly, I did all the work anyway.
